infoschema/builder.go

```diff
@@ -0,0 +1,253 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package infoschema
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/meta"
+	"github.com/pingcap/tidb/meta/autoid"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/table/tables"
+)
+
+// Builder builds a new InfoSchema.
+type Builder struct {
+	is     *infoSchema
+	handle *Handle
+}
+
+// ApplyDiff applies SchemaDiff to the new InfoSchema.
+func (b *Builder) ApplyDiff(m *meta.Meta, diff *model.SchemaDiff) error {
+	b.is.schemaMetaVersion = diff.Version
+	if diff.Type == model.ActionCreateSchema {
+		return b.applyCreateSchema(m, diff)
+	} else if diff.Type == model.ActionDropSchema {
+		b.applyDropSchema(diff.SchemaID)
+		return nil
+	}
+	di, ok := b.is.schemas[diff.SchemaID]
+	if !ok {
+		return ErrDatabaseNotExists
```

Comment:
There must be some serious bug if we go into the branch. Maybe we should log fatal message.